packages/proto-loader/bin/proto-loader-gen-types.ts

```diff
@@ -0,0 +1,745 @@
+  } else {
+    return 0;
+  }
+}
+
+function isNamespaceBase(obj: Protobuf.ReflectionObject): obj is Protobuf.NamespaceBase {
+  return Array.isArray((obj as Protobuf.NamespaceBase).nestedArray);
+}
+
+function stripLeadingPeriod(name: string) {
+  return name.startsWith('.') ? name.substring(1) : name;
+}
+
+function getImportPath(to: Protobuf.Type | Protobuf.Enum | Protobuf.Service): string {
+  /* If the thing we are importing is defined in a message, it is generated in
+   * the same file as that message. */
+  if (to.parent instanceof Protobuf.Type) {
+    return getImportPath(to.parent);
+  }
+  return stripLeadingPeriod(to.fullName).replace(/\./g, '/');
+}
+
+function getPath(to: Protobuf.Type | Protobuf.Enum | Protobuf.Service) {
+  return stripLeadingPeriod(to.fullName).replace(/\./g, '/') + '.ts';
+}
+
+function getPathToRoot(from: Protobuf.NamespaceBase) {
+  const depth = stripLeadingPeriod(from.fullName).split('.').length - 1;
+  if (depth === 0) {
+    return './';
+  }
+  let path = '';
+  for (let i = 0; i < depth; i++) {
+    path += '../';
+  }
+  return path;
+}
+
+function getRelativeImportPath(from: Protobuf.Type | Protobuf.Service, to: Protobuf.Type | Protobuf.Enum | Protobuf.Service) {
+  return getPathToRoot(from) + getImportPath(to);
+}
+
+function getTypeInterfaceName(type: Protobuf.Type | Protobuf.Enum | Protobuf.Service) {
+  return type.fullName.replace(/\./g, '_');
+}
+
+function getImportLine(dependency: Protobuf.Type | Protobuf.Enum | Protobuf.Service, from?: Protobuf.Type | Protobuf.Service) {
+  const filePath = from === undefined ? './' + getImportPath(dependency) : getRelativeImportPath(from, dependency);
+  const typeInterfaceName = getTypeInterfaceName(dependency);
+  let importedTypes: string;
+  /* If the dependenc is defined within a message, it will be generated in that
+   * message's file and exported using its typeInterfaceName. */
+  if (dependency.parent instanceof Protobuf.Type) {
+    if (dependency instanceof Protobuf.Type) {
+      importedTypes = `${typeInterfaceName}, ${typeInterfaceName}__Output`;
+    } else if (dependency instanceof Protobuf.Enum) {
+      importedTypes = `${typeInterfaceName}`;
+    } else if (dependency instanceof Protobuf.Service) {
+      importedTypes = `${typeInterfaceName}Client`;
+    } else {
+      throw new Error('Invalid object passed to getImportLine');
+    }
+  } else {
+    if (dependency instanceof Protobuf.Type) {
+      importedTypes = `${dependency.name} as ${typeInterfaceName}, ${dependency.name}__Output as ${typeInterfaceName}__Output`;
+    } else if (dependency instanceof Protobuf.Enum) {
+      importedTypes = `${dependency.name} as ${typeInterfaceName}`;
+    } else if (dependency instanceof Protobuf.Service) {
+      importedTypes = `${dependency.name}Client as ${typeInterfaceName}Client`;
+    } else {
+      throw new Error('Invalid object passed to getImportLine');
+    }
+  }
+  return `import { ${importedTypes} } from '${filePath}';`

```

Comment:
thoughts on adding an option to output [`import type`](https://devblogs.microsoft.com/typescript/announcing-typescript-3-8/#type-only-imports-exports)?

Reply:
Would be great, I also struggled with it.

Reply:
That does look useful, though I am a little concerned about having an option that is essentially a TypeScript version compatibility switch. I would prefer to always output `export type`, but I am also not sure if it makes sense to assume that the user always has TypeScript 3.8 or higher.

Reply:
We don't really have any current users of the TypeScript generator, so at least that's not a breaking change 🙂 People who use gRPC with TypeScript and would be interested in using/migrating to TypeScript generator are probably already advanced enough to keep their dependencies up to date. Stable version of TypeScript 3.8 has been available for over half a year. Just my two cents 🙂

Reply:
I guess in the worst case we can add a compatibility option later if we need one. OK, I am willing to change the generator to generate `import type` and `export type`.